### PR TITLE
pybind11 type_caster for at::Generator and custom RNG python test

### PR DESCRIPTION
**Pull request by pbelevich**
This PR provides pybind11's `type_caster<at::Generator>` that allows mapping `at::Generator` instance returned from user-defined method to python `torch::Generator`, defined as `THPGenerator ` c++ class. 

This allows 1) defining custom RNG in c++ extension 2) using custom RNG in python code. 

`TestRNGExtension.test_rng` shows how to use custom RNG defined in `rng_extension.cpp`

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35167 Move normal() to DistributionTemplates
* **#34774 pybind11 type_caster for at::Generator and custom RNG python test**

Differential Revision: [D20549451](https://our.internmc.facebook.com/intern/diff/D20549451)

### CODE REVIEW

File: torch/csrc/Generator.cpp [hunk 1]
```diff
@@ -221,3 +221,38 @@ bool THPGenerator_init(PyObject *module)
...
+
+  if (auto obj = pyobj(gen)) {
+    Py_INCREF(obj);
+    return obj;
+  }
+
+  return THPGenerator_NewWithVar((PyTypeObject *)THPGeneratorClass, std::move(gen));
+}
+
+PyObject* THPGenerator_NewWithVar(PyTypeObject* type, Generator gen)
```
Comment (ezyang) [inline torch/csrc/Generator.cpp]: The place you copy pasted this from has a comment talking about what invariant should be upheld before calling this function; that comment should be preserved here!

**Comment by ezyang** [comment]
General comment: I know there is existing Python binding code for Generator, and I don't see that code being adjusted in this PR. So I feel there is definitely something missing here.

**Comment by ezyang** [comment]
Destructor for THPGenerator needs to clear the pyobj field (otherwise it's dangling). Needs more tests about preserving object identity.

File: torch/csrc/Generator.cpp (hunk 1 above)
Comment (ezyang) [reply torch/csrc/Generator.cpp]: don't forget about this comment!

**Review by ezyang — APPROVED** [review]

File: torch/csrc/Generator.cpp (hunk 1 above)
Comment (pbelevich) [reply torch/csrc/Generator.cpp]: I added a comment in the header, but I will add it here as well

**Pull request closed without merging**